### PR TITLE
App Controlled Execution Config

### PR DESCRIPTION
**Pull request by nibanks**
## Description

Refactors the MsQuic API to allow for the application to explicitly configure the execution of MsQuic internals. This is the next step to allow the app to fully control execution.

## Testing

Existing CI

## Documentation

to-do

### CODE REVIEW

File: src/perf/lib/SecNetPerfMain.cpp [hunk 1]
```diff
@@ -96,9 +92,10 @@ PrintHelp(
         "  -cc:<algo>                  Congestion control algorithm to use.\n"
         "\n"
         "Client: secnetperf -TestName:<Throughput|RPS|HPS> [options]\n"
-#ifndef _KERNEL_MODE
         "Both:\n"
-        "  -cpu:<cpu_index>            Specify the processor(s) for the datapath to use.\n"
+        "  -cpu:<cpu_index>            Specify the processor(s) to use.\n"
```
Comment (nibanks) [inline src/perf/lib/SecNetPerfMain.cpp:96]: Specify `sleepthresh`

File: src/core/registration.c [hunk 2]
```diff
@@ -58,39 +54,15 @@ MsQuicRegistrationOpen(
...
-                    LibraryErrorStatus,
-                    "[ lib] ERROR, %u, %s.",
-                    Status,
-                    "CxPlatDataPathInitialize");
-                goto Error;
-            }
-            QuicTraceEvent(
-                DataPathInitialized,
-                "[data] Initialized, DatapathFeatures=%u",
-                CxPlatDataPathGetSupportedFeatures(MsQuicLib.Datapath));
```
Comment (csujedihy) [inline src/core/registration.c:89]: This refactor seems unnecessary to me. QuicLibraryEnsureExecutionContext is essentially just initialize datapath.

File: src/platform/platform_worker.c [hunk 3]
```diff
@@ -76,49 +62,74 @@ typedef struct QUIC_CACHEALIGN CXPLAT_WORKER {
...
 #pragma warning(disable:6386) // SAL is confused about the worker size
 BOOLEAN
-CxPlatWorkersInit(
-    void
+CxPlatWorkersLazyStart(
+    _In_opt_ QUIC_EXECUTION_CONFIG* Config
     )
 {
-    CxPlatWorkerCount = CxPlatProcActiveCount(); // TODO - use max instead?
+    CxPlatLockAcquire(&CxPlatWorkerLock);
```
Comment (csujedihy) [inline src/platform/platform_worker.c:106]: >CxPlatWorkerLock [](http://example.com/codeflow?start=23&length=16)

Is this lock necessary? It's only used in CxPlatWorkersLazyStart. I can't see how CxPlatWorkersLazyStart can be called from different threads.

File: src/core/registration.c (hunk 2 above)
Comment (nibanks) [reply src/core/registration.c:89]: As it is now, I'd agree with you. But in an effort to break this PR up some, I haven't made the QUIC layer changes to only run on the configured CPUs. That's the next step. With that work, a lot more will change in library.c around PartitionCount, so more will be added then.

File: src/platform/platform_worker.c (hunk 3 above)
Comment (nibanks) [reply src/platform/platform_worker.c:106]: The platform layer is treated as a semi-public API, that could possibly be used my multiple consumers. Indeed, if you compile our tools with static linking, then the platform does end up getting shared by MsQuic and the tool itself. So just to be extra safe, we have synchronization.

File: scripts/RemoteTests.json [hunk 4]
```diff
@@ -279,7 +279,7 @@
                 "Tls": ["openssl"],
                 "Arch": ["x64", "arm"],
                 "Exe": "secnetperf",
-                "Arguments": "-test:HPS -target:$RemoteAddress"
+                "Arguments": "-exec:lowlat -test:HPS -target:$RemoteAddress"
```
Comment (nibanks) [inline scripts/RemoteTests.json:282]: Maybe `maxtput` for this one?

File: scripts/RemoteTests.json [hunk 5]
```diff
@@ -172,7 +172,7 @@
                 "Tls": ["schannel", "openssl"],
                 "Arch": ["x64", "x86", "arm", "arm64"],
                 "Exe": "secnetperf",
-                "Arguments": "-test:RPS -target:$RemoteAddress"
+                "Arguments": "-exec:lowlat -test:RPS -target:$RemoteAddress"
```
Comment (nibanks) [inline scripts/RemoteTests.json:175]: RPS tests both throughput and latency, depending on the test. So I am unsure of what to use? Both?

File: src/inc/msquic.h [hunk 6]
```diff
@@ -232,6 +232,27 @@ typedef enum QUIC_DATAGRAM_SEND_STATE {
...
+
+DEFINE_ENUM_FLAG_OPERATORS(QUIC_EXECUTION_CONFIG_FLAGS)
+
+//
+// A custom configuration for thread execution in QUIC.
+//
+typedef struct QUIC_EXECUTION_CONFIG {
+
+    QUIC_EXECUTION_CONFIG_FLAGS Flags;
+    uint32_t SleepTimeoutUs;            // Number of microseconds of idleness before sleeping a datapath thread.
```
Comment (nibanks) [inline src/inc/msquic.h:247]: PollingIdleTimeoutUs

Time before a polling thread, with no work to do, sleeps.

**Review by maolson-msft — APPROVED** [review]

File: src/core/worker.c [hunk 7]
```diff
@@ -101,29 +95,36 @@ QuicWorkerInitialize(
...
-    if (QUIC_FAILED(Status)) {
-        QuicTraceEvent(
-            WorkerErrorStatus,
-            "[wrkr][%p] ERROR, %u, %s.",
-            Worker,
-            Status,
-            "CxPlatThreadCreate");
-        goto Error;
+#ifndef _KERNEL_MODE // Not supported on kernel mode
+    if (ExecProfile != QUIC_EXECUTION_PROFILE_TYPE_MAX_THROUGHPUT) {
```
Comment (ami-GS) [inline src/core/worker.c:99]: Does max throughput related flag works without splitpartitioning?

Comment (nibanks) [reply src/core/worker.c:99]: Yes. Max throughput just means separate threads for UDP and QUIC now.

**Pull request merged**